javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1014,6 +1043,201 @@ object Formatter {
     )
   }
 
+  /** Converts a `Double` into a `Decimal` that has as few digits as possible
+   *  while still uniquely identifying `x`.
+   *
+   *  We do this by converting the absolute value of the number into a string
+   *  using its built-in `toString()` conversion. By ECMAScript's spec, this
+   *  yields a decimal representation with as few significant digits as
+   *  possible, although it can be in fixed notation or in computerized
+   *  scientific notation.
+   *
+   *  We then parse that string to recover the integer part, the factional part
+   *  and the exponent; the latter two being optional.
+   *
+   *  From the parts, we construct a `Decimal`, making sure to create one that
+   *  does not have leading 0's (as it is forbidden by `Decimal`'s invariants).
+   */
+  private def numberToDecimal(x: Double): Decimal = {
+    if (x == 0.0) {
+      val negative = 1.0 / x < 0.0
+      Decimal.zero(negative)
+    } else {
+      val negative = x < 0.0
+      val s = JDouble.toString(if (negative) -x else x)
+
+      val ePos = s.indexOf('e')
+      val e =
+        if (ePos < 0) 0
+        else js.Dynamic.global.parseInt(s.substring(ePos + 1)).asInstanceOf[Int]
+      val significandEnd = if (ePos < 0) s.length() else ePos
+
+      val dotPos = s.indexOf('.')
+      if (dotPos < 0) {
+        // No '.'; there cannot be leading 0's (x == 0.0 was handled before)
+        val unscaledValue = s.substring(0, significandEnd)
+        val scale = -e
+        new Decimal(negative, unscaledValue, scale)
+      } else {
+        // There is a '.'; there can be leading 0's, which we must remove
+        val digits = s.substring(0, dotPos) + s.substring(dotPos + 1, significandEnd)
+        val digitsLen = digits.length()
+        var i = 0
+        while (i < digitsLen && digits.charAt(i) == '0')
+          i += 1
+        val unscaledValue = digits.substring(i)
+        val scale = -e + (significandEnd - (dotPos + 1))
+        new Decimal(negative, unscaledValue, scale)
+      }
+    }
+  }
+
+  /** A decimal representation of a number.
+   *
+   *  An instance of this class represents the number whose absolute value is
+   *  `unscaledValue × 10^(-scale)`, and that is negative iff `negative` is
+   *  true.
+   *
+   *  The `unscaledValue` is stored as a String of decimal digits, i.e.,
+   *  characters in the range ['0', '9'], expressed in base 10. Leading 0's are
+   *  *not* valid.
+   *
+   *  As an exception, a zero value is represented by an `unscaledValue` of
+   *  `"0"`. The scale of zero value is always 0.
+   *
+   *  `Decimal` is similar to `BigDecimal`, with some differences:
+   *
+   *  - `Decimal` distinguishes +0 from -0.
+   *  - The unscaled value of `Decimal` is stored in base 10.
+   *
+   *  The methods it exposes have the same meaning as for BigDecimal, with the
+   *  only rounding mode being HALF_UP.
+   */
+  private final class Decimal(val negative: Boolean, val unscaledValue: String,
+      val scale: Int) {
```

Comment:
There is one thing I do not understand here: Don't we have a degree of freedom too much? Wouldn't it be better to disallow *trailing* zeroes as well in `unscaledValue`?

Reply:
There is a degree of freedom, absolutely, and it is intentional. It is very helpful in the `decimalNotation`, for example, to be able to set the scale of Decimal to an arbitrary value, even if that requires to add trailing zeros. It is also helpful in `numberToDecimal`, where we do not have to deal with removing trailing zeros (only to perhaps add them back in the notation function). In the next PR, it is also helpful because trailing zeros are not forbidden in `BigDecimal`, which means that the conversion is easier if we accept trailing zeros as well.

Also, looking at it in the other direction: forbidding trailing zeros would not make anything simpler: there is no code path that we could simplify with the knowledge that there are no trailing zeros.

Reply:
I still do not really have a good intuition of where the abstraction boundary of a `Decimal` lies :-/ I have added a bunch of comments of what I'm thinking, but I'm not super convinced that they are in the right direction.

Maybe my confusion comes from the different meaning of `precision` based on context (but IIUC, this is out of our hands, it's essentially `Formatter` spec).

I'll try to give this another look later today.

Reply:
> I still do not really have a good intuition of where the abstraction boundary of a `Decimal` lies :-/

Perhaps it helps *not* to think too much about it as an *abstraction*. In this context, it's not more of an abstraction over the digits/scale than `Double` is an abstraction over mantissa/exponent. We need to break the abstraction to properly format Doubles (in `formatHexFloatingPoint`) as much as `Decimal`s (with this PR). Consider it more as a mere struct that encapsulates related values, but that doesn't give a real abstraction; the same way we consider Double as a struct of (sign, mantissa, exponent).

> Maybe my confusion comes from the different meaning of `precision` based on context (but IIUC, this is out of our hands, it's essentially `Formatter` spec).

Perhaps. Just as a recap:
* For `%e`, the precision is the number of digits after the `.`, knowing that only exactly 1 digit will be before; it therefore corresponds to `unscaledValue.length() - 1` of a `Decimal`.
* For `%f`, the precision is the number of digits *after* the `.`; it therefore directly corresponds the notion of *scale* of a `Decimal`.
* For `%g`, the precision is the total number of significant digits, whether they appear on the left or on the right of a `.`. It therefore corresponds to `unscaledValue.length()` of a `Decimal`.

`setScale` can be used to fix the scale. `round` can be used to fix the `unscaledValue.length` (which is called `precision` in `Decimal` because that's its name in `BigDecimal`).

Reply:
Yeah, I think that makes sense. So let's keep it as is.